source/common/network/lc_trie.h

```diff
@@ -0,0 +1,567 @@
+      uint32_t count;
+      do {
+        ++branch;
+
+        // Check if the current branch factor with the fill factor can contain all of the nodes
+        // in the current range or if the current branching factor is larger than the
+        // IP address_size.
+        if (n < fill_factor_ * (1 << branch) ||
+            static_cast<uint32_t>(compute.prefix_ + branch) > address_size) {
+          break;
+        }
+
+        // Start by checking the bit patterns at ip_prefixes_[first] through
+        // ip_prefixes_[first + n-1].
+        index = first;
+        // Pattern to search for.
+        uint32_t pattern = 0;
+        // Number of patterns found while looping through the list.
+        count = 0;
+
+        // Search for all patterns(values) within 1<<branch.
+        while (pattern < static_cast<uint32_t>(1 << branch)) {
+          bool pattern_found = false;
+          // Keep on looping until either all nodes in the range have been visited or
+          // an IP prefix doesn't match the pattern.
+          while (index < first + n &&
+                 static_cast<uint32_t>(extractBits<IpType, address_size>(
+                     compute.prefix_, branch, ip_prefixes_[index].ip_)) == pattern) {
+            ++index;
+            pattern_found = true;
+          }
+
+          if (pattern_found) {
+            ++count;
+          }
+          ++pattern;
+        }
+        // Stop iterating once the size of the branch (with the fill factor ratio) is able to
+        // contain all of the prefixes within the current range of ip_prefixes_[first] to
+        // ip_prefixes_[first+n-1].
+      } while (count >= fill_factor_ * (1 << branch));
+
+      // Set the branching factor.
+      compute.branch_ = branch - 1;
+      return compute;
+    }
+
+    /**
+     * Recursively build a trie for IP prefixes from position 'first' to 'first+n-1'.
+     * @param prefix supplies the prefix to ignore when building the sub-trie.
+     * @param first supplies the index into ip_prefixes_ for this sub-trie.
+     * @param n suppplies the number of entries for the sub-trie.
+     * @param position supplies the root for this sub-trie.
+     * @param next_free_index supplies the next available index in the trie_.
+     */
+    void buildRecursive(uint32_t prefix, uint32_t first, uint32_t n, uint32_t position,
+                        uint32_t& next_free_index) {
+      // Setting a leaf, the branch and skip are 0.
+      if (n == 1) {
+        // There is no way to predictably determine the number of trie nodes required to build a
+        // LC-Trie. If while building the trie the position that is being set exceeds the maximum
+        // number of supported entries, throw an Envoy Exception instead of letting an out_of_range
+        // exception be thrown.
+        // TODO(ccaraman): Add a test case.
+        if (position > MAXIMUM_TRIE_NODES) {
+          throw EnvoyException(fmt::format("The number of internal nodes required for the LC-Trie "
```

Comment:
I know this is not the responsibility of the throwing code. But given we own both the throwing code and the code that receives the exception, I want to make sure this is caught at config time, which I am 99% sure it does.

Reply:
Yes, this will get thrown during config parsing  The call chain is LcTrie()->LcTrieInternal<>-> build()->buildRecursive